### PR TITLE
Feature notes

### PR DESCRIPTION
**Pull request by natlibfi-jonollil**

### CODE REVIEW

File: src/mongoNotes.js [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    * @param {*} appsList
+    * @returns {Boolean}
+    */
+    function isValidContext(appsList) {
+      return appsList.every((app) => validContext.includes(app));
+    }
+
+    /**
+     * Validates if message text is not empty and is String
+     * @param {*} messageText
```
Comment (natlibfi-jonollil) [inline src/mongoNotes.js:151]: Incoming param should be string?

File: src/mongoNotes.js [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+     * @param {*} property
+     * @returns {Boolean}
+     */
+    function isBoolean(property) {
+      return typeof property === 'boolean';
+    }
+
+    /**
+    * Validates if every app in list is valid context
+    * @param {*} appsList
```
Comment (natlibfi-jonollil) [inline src/mongoNotes.js:142]: Incoming param should be array of strings? `[String]`

File: src/mongoNotes.js [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+     * @param {*} noteItem
+     * @returns {Boolean}
+     */
+    function isNotObject(noteItem) {
+      return typeof noteItem !== 'object' || Object.keys(noteItem).length === 0 || Object.getPrototypeOf(noteItem) !== Object.prototype;
+    }
+
+    /**
+     * Validates if property is Boolean
+     * @param {*} property
```
Comment (natlibfi-jonollil) [inline src/mongoNotes.js:133]: Incoming param here should be boolean?

File: src/mongoNotes.js [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    function validate(property, validator) {
+      if (property === undefined) {
+        return undefined;
+      }
+      return validator(property) ? property : undefined;
+    }
+
+    /**
+     * Checks if noteItem is Object
+     * @param {*} noteItem
```
Comment (natlibfi-jonollil) [inline src/mongoNotes.js:124]: Incoming param should be string?

File: src/mongoNotes.js [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    throw new ApiError(httpStatus.INTERNAL_SERVER_ERROR, `MongoNotes had error while adding a note ${noteItem}`);
+
+
+    /**
+     * Validate property with validator function
+     * @param {*} property
+     * @param {function} validator
+     * @returns {(*|undefined)}
+     */
+    function validate(property, validator) {
```
Comment (natlibfi-jonollil) [inline src/mongoNotes.js:115]: Is this validate wrapper nessessary? Could those validation functions be free? And maybe even in utils.js or validate.js file?

Comment (natlibfi-kmskuusi) [reply src/mongoNotes.js]: Notes validation functions and variables moved to validateNote.js and updated import/export usage as such.

File: src/mongoNotes.js (hunk 1 above)
Comment (natlibfi-kmskuusi) [reply src/mongoNotes.js]: Param comments updated

File: src/mongoNotes.js (hunk 2 above)
Comment (natlibfi-kmskuusi) [reply src/mongoNotes.js]: Param comments updated

File: src/mongoNotes.js (hunk 3 above)
Comment (natlibfi-kmskuusi) [reply src/mongoNotes.js]: Param comments updated

File: src/mongoNotes.js (hunk 4 above)
Comment (natlibfi-kmskuusi) [reply src/mongoNotes.js]: Expected object. Param comments updated

File: src/mongoNotes.js (hunk 5 above)
Comment (natlibfi-kmskuusi) [reply src/mongoNotes.js]: Wrapper is now removed and only data validator functions are used. Not valid data handling is in the usage level not baked in to the wrapper. Easier to set not valid default values if so needed.

**Review by natlibfi-kmskuusi — APPROVED** [review]
Changes have been made as per request. Collaboration changes seems to be as required. Tests pass. I see no issue here

**Review by natlibfi-ptalosel — APPROVED** [review]

**Comment by natlibfi-kmskuusi** [comment]
Converted to draft. Re-checking additional field requirements for ie. linkButton url.

**Comment by natlibfi-kmskuusi** [comment]
Latest changes by me so requesting other people to re-review.
Url field added and tests updated. Merged new-ui-rest to branch to resolve file conflicts github listed.

**Comment by natlibfi-kmskuusi** [comment]
Theres issue with ES modules type with test/testFixtures/notes.spec.js. @natlibfi-ptalosel told to take care of it. Setting myself re-review also when fix arrives.

File: package.json [hunk 6]
```diff
@@ -23,37 +23,32 @@
     "node": ">=20"
```
Comment (natlibfi-kmskuusi) [inline package.json:23]: Seems like lower node versions (tested on 18.12.1) and minor 20 versions (tested on 20.6.1) throw error
`TypeError: The "path" argument must be of type string. Received undefined`
when running tests.

This however is not an issue on 20 lts (20.11.1) and  onwards.

Should we specify just to make sure ? 

`"node": ">=20.11.1"`

Most likely people tend to install lts version of major required ones but do we want to make sure ?

**Review by natlibfi-kmskuusi — APPROVED** [review]
Since latests "self review" changes make sense.
Module issue cleared.
Running dev tests check out.
Node tests clear.

**Review by natlibfi-ptalosel — APPROVED** [review]

**Pull request merged**